zerozeroreadme/zerozeroreadme/__init__.py

```diff
@@ -285,6 +300,80 @@ def is_keep_comments(self, testing: str) -> bool:
                     return True
         return False
 
+    def update_from_preflight(self, pf: PreflightResponse) -> bool:
+        """Update ZZRM variables from Preflight response."""
+        if not self.toplevels:
+            for tlf in pf.detected_toplevel_files:
+                if tlf.filename in self.sources:
+                    # it cannot be a toplevel usage file, since we check
+                    # for toplevel files above, so it must be either ignore etc
+                    # Obey this indirection and ignore the PreFlight, we might
+                    # have been wrong here
+                    pass
+                else:
+                    uf = UserFile(filename=tlf.filename, usage=FileUsageType.toplevel)
+                    self.sources[tlf.filename] = uf
+        # if we still not have any toplevel files, because all files found by
+        # Preflight have been marked as ignored/append etc, then bail out
+        if not self.toplevels:
+            return False
+        # Now we are sure we have toplevel files set.
+        # If no compiler is selected, select it based on the first toplevel file
+        if self.process.compiler is None:
+            self.process.compiler = CompilerSpec(
+                engine=EngineType.unknown, lang=LanguageType.unknown, output=OutputType.unknown
+            )
+        if not self.process.compiler.is_determined:
+            first_tex = self.toplevels[0]
+            # search for first_tex in the detected toplevel files
+            found_tlp: ToplevelFile | None = None
+            for tlp in pf.detected_toplevel_files:
+                if tlp.filename == first_tex:
+                    found_tlp = tlp
+                    break
+            if not found_tlp:
+                # Couldn't find selected file in list of toplevel files
+                return False
+            if self.process.compiler.engine == EngineType.unknown:
+                if found_tlp.process.compiler.engine == EngineType.unknown:
+                    self.process.compiler.engine = DEFAULT_ENGINE_TYPE
+                else:
+                    self.process.compiler.engine = found_tlp.process.compiler.engine
+            if self.process.compiler.lang == LanguageType.unknown:
+                if found_tlp.process.lang == LanguageType.unknown:
+                    self.process.compiler.lang = DEFAULT_LANGUAGE_TYPE
+                else:
+                    self.process.compiler.lang = found_tlp.process.compiler.lang
+            if self.process.compiler.output == OutputType.unknown:
+                if found_tlp.process.output == OutputType.unknown:
```

Comment:
Is  `found_tlp.process.output` missing the `.compiler` sub-component here?

Namely: `found_tlp.process.compiler.output`

Same question for the `process.lang` just above.
Basing my question on the [MainProcessSpec](https://github.com/arXiv/submission-tools/blob/60c51a1f760c15a5ae362903b5239b9a6bb7a3df/preflight_parser/preflight_parser/__init__.py#L268) definition, apologies in advance if I am misunderstanding the details here.

Reply:
Thanks, good catch, completely right. Fixed now.